src/v/kafka/server/handlers/fetch.cc

```diff
@@ -281,6 +282,74 @@ make_partition_response_error(model::partition_id p_id, error_code error) {
     };
 }
 
+uint32_t
+control_record_size(int64_t ts_delta, int32_t offset_delta, const iobuf& key) {
+    static constexpr size_t zero_vint_size = vint::vint_size(0);
+    return sizeof(model::record_attributes::type) // attributes
+           + vint::vint_size(ts_delta)            // timestamp delta
+           + vint::vint_size(offset_delta)        // offset_delta
+           + vint::vint_size(key.size_bytes())    // key size
+           + key.size_bytes()                     // key payload
+           + zero_vint_size                       // value size
+           + zero_vint_size;                      // headers size
+}
+
+iobuf make_control_record_batch_key() {
+    iobuf b;
+    response_writer w(b);
+    /**
+     * control record batch schema:
+     *   [version, type]
+     */
+    w.write(model::current_control_record_version);
+    w.write(model::control_record_type::unknown);
+    return b;
+}
+
+/**
+ * here we make sure that our internal control batches are correctly adapted
+ * for Kafka clients
+ */
+model::record_batch adapt_fetch_batch(model::record_batch&& batch) {
+    // pass through data batches
+    if (likely(batch.header().type == raft::data_batch_type)) {
+        return std::move(batch);
```

Comment:
I think you're inhibiting NRVO with the `std::move`, is this not a warning?

Reply:
Without `std::move` it would call copy constructor, `batch` is `&&` rvalue ref

Reply:
Sorry, not sure what I was thinking.